### DISCUSSION TITLE
Opaque

### DISCUSSION THREAD

**Discussion by serkosen** [zeroc-ice/ice#1410] [Q&A]
Hi,

I need to store opaque Ice objects as, say, cache between sessions possibly serialized or to relay. I use Ice 3.7.8.. It seems to work for encoding v 1.1 similar to UnknownSlicedValue and with help of ValueFactory for objects of any type - "":

```
void
Ice::UnknownSlicedValue::_iceWrite(Ice::OutputStream* ostr) const
{
    ostr->startValue(_slicedData);
    ostr->endValue();
}

void
Ice::UnknownSlicedValue::_iceRead(Ice::InputStream* istr)
{
    istr->startValue();
    _slicedData = istr->endValue(true);
}

```
I unsuccessfully tried to make the above work for legacy encoding v 1.0.  What is the right way to do it?

Thanks a lot,
Sergei Kosenko

**Comment by bernardnormier** [comment]
Hi Sergei,

Can you describe your use-case at a higher level? What would your Slice definitions look like?

> I need to store opaque Ice objects

By Ice objects, you mean instances of Slice classes?
Can you choose how to represent your data in Slice, or is this an existing application and you have to use Slice classes?

Note that the "preservation" of unknown derived "slices" of a class instance during decoding was introduced in encoding 1.1. Encoding 1.0 does not support this feature. Are you sure you need to support encoding 1.0?

Best regards,
Bernard

**Comment by serkosen** [comment]
Hi Bernard,

Thanks for quick reply. We upgrade Ice from 3.4.x to 3.7.8 version. We stage our system upgrade few apps at a time. Version 1.0 encoding has been chosen for backward compatibility; all will be upgraded to version 1.1 encoding once all app are upgraded.
By opaque I mean objects of unknown for the cache server type that are sent there and stored as key <-> blob sequence of slices pairs and rendered to other applications upon request. 

Below works in Ice 3.4.x :
```
//slice snip - I may be able to play with how dictionary of key<->objects are sent
dictionary<string, Object> ObjectDict;
interface SubjectPublisher {
    idempotent void publish( string name, Object data );
    idempotent void publish_many( ObjectDict objects );
};
//cpp snip
...
void OpaqueSlicedObject::__read( IceInternal::BasicStream *is, bool rid ) 
{   
    do {
        if( rid ) {
            std::string typeName;
            is->readTypeId( typeName ); 
            this->m_typeData.push_back( typeName );
        }
        Ice::Int sz = 0;
        is->read( sz );
        if( size_t( sz ) < sizeof( sz ) )
            throw Ice::UnmarshalOutOfBoundsException( __FILE__, __LINE__ );
        sz -= sizeof( sz );
        if( sz )
            is->readBlob( this->m_typeData.back().m_data, sz );
        if( this->m_typeData.back().m_typeName == Ice::Object::ice_staticId() )
            return;
        rid = true;
    } while ( true ) 
}
```
Above used to read unknown object into sequence of slices as blobs, that then got sent back upon request written into the stream in same order. 
Hope this better explains my problem.

Thanks a lot,
Sergei Kosenko

**Comment by bernardnormier** [comment]
Hi Sergei,

So with OpaqueSlicedObject, you implemented your own "unknown/opaque class instance preservation" with Ice 3.4. I just double-checked there is no such class in Ice 3.4. Pretty advanced.

The 1.0 encoding for classes did not change (obviously) between Ice 3.4 and Ice 3.7, so it seems to me you should:
- as long as you use the 1.0 encoding, keep the same approach, with your own custom implementation of "Object" in C++. You can port your OpaqueSlicedObject to Ice 3.7.x (this implementation would remain encoding 1.0-only).
- when you can drop the 1.0 encoding and switch to the 1.1 encoding, replace OpaqueSlicedObject by the Ice-provided `UnknownSlicedValue`. Or maybe keep your OpaqueSlicedObject API and reimplement it using UnknownSlicedValue?

Ice 3.7 uses the term "Value" to mean instance of Slice class passed by value through a Slice operation. In 3.4, it was called "Object". 

The class `Ice::UnknownSlicedValue` does not support encoding 1.0.

Then, with encoding 1.1, make sure you use the "sliced format" when encoding your Slice class instances (you did not show how you're encoding these instances). The default class format in Ice 3.7 is the "compact format". See https://doc.zeroc.com/ice/3.7/client-server-features/slicing-values-and-exceptions#id-.SlicingValuesandExceptionsv3.7-SliceFormats

Best regards,
Bernard
